core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -1292,6 +1292,12 @@ class ConnectionQuotas(config: KafkaConfig, time: Time, metrics: Metrics) extend
     counts.synchronized {
       val startThrottleTimeMs = time.milliseconds
       val throttleTimeMs = math.max(recordConnectionAndGetThrottleTimeMs(listenerName, startThrottleTimeMs), 0)
+      if (throttleTimeMs > 0) {
+        // record throttle time due to hitting connection rate limit
+        // connection could be throttled longer if the limit of the number of active connections is reached as well
+        maxConnectionsPerListener.get(listenerName)
+          .foreach(_.connectionRateThrottleSensor.record(throttleTimeMs.toDouble, startThrottleTimeMs))
```

Comment:
It is a bit annoying that we have to lookup the `ListenerConnectionQuota` twice. Once in `recordConnectionAndGetThrottleTimeMs` and once here. I wonder if we could look it up once to record the rate and to record the throttle time.

I am thinking about the following to be more concrete. We could add two methods in `ListenerConnectionQuota`: 1) `record` and 2) `recordThrottleTime`. They would encapsulate the logic to respectively record the rate (for the listener and the broker like we do now) and the throttle time. So here we could look up the `ListenerConnectionQuota` record and get the throttle time, and record the throttle time if > 0.

That could improve the readability a bit. I am not sure that it would make a difference from a performance point of view. WDYT?

Reply:
That code seemed readable to me, but perhaps looking up `ListenerConnectionQuota` is better. I made a change to lookup once, please take a look.